orangecontrib/text/topics/topics.py

```diff
@@ -0,0 +1,122 @@
+from gensim import matutils
+
+from Orange.data import StringVariable, ContinuousVariable, Domain
+from Orange.data.table import Table, np
+from orangecontrib.text.corpus import Corpus
+
+
+def chunks(iterable, chunk_size):
+    chunk = []
+    for item in iterable:
+        chunk.append(item)
+        if len(chunk) == chunk_size:
+            yield chunk
+            chunk = []
+    if chunk:
+        yield chunk
+
+
+MAX_WORDS = 1000
+
+
+class Topics(Table):
+    """ Dummy wrapper for Table so signals can distinguish Topics from Data.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """ Bypass Table.__new__. """
+        return object.__new__(Topics)
+
+
+class GensimWrapper:
+    name = NotImplemented
+    Model = NotImplemented
+
+    def fit(self, corpus, progress_callback=None):
+        """ Train the model with the corpus.
+
+        Args:
+            corpus (Corpus): A corpus to lear topics from.
+        """
+        if not len(corpus.dictionary):
+            return None
+        self.reset_model(corpus)
+        self.update(corpus, progress_callback)
+        self.topic_names = ['Topic{} ({})'.format(i, ', '.join(words))
+                            for i, words in enumerate(self._topics_words(3), 1)]
+
+    def reset_model(self, corpus):
+        self.model = self.Model(id2word=corpus.ngrams_dictionary)
+
+    def update(self, corpus, progress_callback=None):
+        gcorpus = matutils.Sparse2Corpus(corpus.ngrams_matrix.T)
+
+        for i, chunk in enumerate(chunks(gcorpus, np.ceil(len(corpus) / 100))):
+            self.model.update(chunk)
+            if progress_callback:
+                progress_callback(i)
+
+    def transform(self, corpus):
+        """ Create a table with topics representation. """
+        topics = self.model[matutils.Sparse2Corpus(corpus.ngrams_matrix.T)]
+        matrix = matutils.corpus2dense(topics,
+                                       num_terms=self.num_topics).T
+
+        # Generate the new table.
```

Comment:
I know this is probably some legacy code but we should change this to use `corpus.extend_attributes`. The current implementation overwrites previous potential  attributes on input.